File: src/harness/agent.py
from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from typing import Any, Callable, Dict, List

from .error_policy import ErrorPolicy
from .llm import BaseLLM
from .logger import TrajectoryLogger
from .memory import MemoryManager
from .schema import SchemaError, parse_llm_action
from .snapshot import SnapshotStore
from .stop_controller import StopController, StopState
from .tools import ToolDispatcher
from .types import RunResult, ToolExecutionResult, TurnRecord


@dataclass
class RunConfig:
    """Configuration for a single agent run.

    Attributes:
        max_steps: Hard upper limit on the number of turns before the run is
            terminated with ``stop_reason="max_steps_reached"``.
        log_dir: Directory where JSONL trajectory logs are written.
        max_history_turns: Number of most-recent turns to include verbatim in
            the working memory sent to the LLM.
        schema_retry_limit: How many times to retry the LLM in the same turn
            when its output fails schema validation. ``1`` means one retry
            attempt (two LLM calls total) before the turn is marked as a
            schema error.
        max_budget: Optional abstract budget counter. Each LLM call and each
            tool attempt increments the counter by 1. The run stops with
            ``stop_reason="max_budget_reached"`` when the counter exceeds
            this value.
        max_failures: Stop the run with ``stop_reason="max_failures_reached"``
            after this many cumulative failures (schema errors + tool errors).
            ``None`` disables the check.
        tool_retry_limit: Maximum number of automatic retries for tool calls
            that raise :class:`~harness.tools.RetryableToolError`. ``0``
            means no retries.
        snapshot_dir: Directory for per-turn snapshot JSON files. Falls back
            to ``log_dir`` when ``None``.
        dangerous_tools: Tool names whose identical calls are blocked after the
            first execution to prevent unintended side-effect repetition.
        goal_reached_token: When set, a final response that contains this
            token (substring match) changes the stop reason to
            ``"goal_reached"``.
        allowed_write_roots: Absolute path prefixes that ``write_text_file``
            is permitted to write into. An empty tuple disables the tool.
    """

    max_steps: int = 8
    log_dir: str = "logs"
    max_history_turns: int = 8
    schema_retry_limit: int = 1
    max_budget: int | None = None
    max_failures: int | None = 3
    tool_retry_limit: int = 0
    snapshot_dir: str | None = None
    dangerous_tools: tuple[str, ...] = ()
    goal_reached_token: str | None = None
    allowed_write_roots: tuple[str, ...] = ()
    project_root: str = ""
    allow_bash: bool = True
    max_tokens_budget: int | None = None
    trust_level: str = "ask"


class HarnessAgent:
    """Single-agent execution harness.

    Orchestrates the ``memory → llm → tool/final → observe → next turn`` loop
    with strict schema validation, reliability guardrails, and trajectory
    logging.

    Args:
        llm: Any :class:`~harness.llm.BaseLLM` implementation that produces
            structured action dicts.
        config: Runtime configuration. Defaults to :class:`RunConfig` with
            all default values when ``None``.
    """

    def __init__(self, llm: BaseLLM, config: RunConfig | None = None) -> None:
        self.llm = llm
        self.config = config or RunConfig()
        self.tools = ToolDispatcher(allowed_write_roots=list(self.config.allowed_write_roots))
        if self.config.project_root:
            from .tools import register_coding_tools

            register_coding_tools(self.tools, allow_bash=self.config.allow_bash)
        if hasattr(self.llm, "set_tool_schemas"):
            self.llm.set_tool_schemas(self.tools.get_tool_schemas())
        self.memory = MemoryManager(max_history_turns=self.config.max_history_turns)
        self.error_policy = ErrorPolicy(tool_retry_limit=self.config.tool_retry_limit)
        self.snapshot_store = SnapshotStore(self.config.snapshot_dir or self.config.log_dir)
        self.stop_controller = StopController(
            max_budget=self.config.max_budget,
            max_failures=self.config.max_failures,
            goal_reached_token=self.config.goal_reached_token,
        )

    def run(
        self,
        goal: str,
        context: Dict[str, Any] | None = None,
        *,
        resume_from: str | None = None,
        on_turn: Callable[["TurnRecord"], None] | None = None,
        on_approve: Callable[[str, str, Dict[str, Any]], bool] | None = None,
        on_token: Callable[[str], None] | None = None,
    ) -> RunResult:
        """Run the agent loop until a stop condition is met.

        Args:
            goal: Natural-language task description passed to the LLM each turn.
            context: Optional key/value pairs injected into working memory.
            resume_from: Path to a snapshot JSON file. When provided, the run
                continues from the saved state rather than starting fresh.
            on_turn: Optional callback invoked after each turn is recorded.
                Receives the completed :class:`~harness.types.TurnRecord`.
                Useful for streaming live output to a UI.
            on_token: Optional callback for streaming token output. Each
                token string is passed as it arrives from the LLM.

        Returns:
            A :class:`~harness.types.RunResult` with the final response, full
            turn history, stop reason, log path, and latest snapshot path.
        """
        self.llm.on_token = on_token
        state = self._load_state(goal=goal, context=context or {}, resume_from=resume_from)
        goal = state["goal"]
        context = state["context"]
        turns = state["turns"]
        self.memory.summary = state["summary"]
        logger = TrajectoryLogger(self.config.log_dir)
        stop_reason = "max_steps_reached"
        snapshot_path = state.get("snapshot_path")
        dangerous_sigs: List[str] = state["dangerous_tool_signatures"]
        runtime_state = StopState(
            budget_used=state["budget_used"],
            failure_count=state["failure_count"],
        )

        total_tokens = 0

        for _ in range(self.config.max_steps):
            stop_reason = self.stop_controller.check_before_turn(runtime_state) or stop_reason
            if stop_reason == "max_budget_reached":
                break

            turn_idx = len(turns) + 1
            working_memory = self.memory.build_working_memory(
                goal=goal, context=context, turns=turns
            )
            action_result = self._generate_action_with_schema_retry(working_memory)

            # Track token usage from LLM responses
            raw = action_result.get("llm_raw_output")
            if isinstance(raw, dict) and "_usage" in raw:
                total_tokens += raw["_usage"].get("total_tokens", 0)
            if self.config.max_tokens_budget and total_tokens > self.config.max_tokens_budget:
                stop_reason = "token_budget_exceeded"
                break
            runtime_state.budget_used += len(action_result["schema_errors"]) + 1
            if not action_result["ok"]:
                runtime_state.failure_count += 1
                stop_reason = (
                    self.stop_controller.check_after_failure(runtime_state) or "schema_error"
                )
                record = TurnRecord(
                    turn=turn_idx,
                    goal=goal,
                    working_memory=working_memory,
                    llm_raw_output=action_result["llm_raw_output"],
                    llm_action={
                        "type": "schema_error",
                        "error": action_result["error"],
                        "schema_errors": action_result["schema_errors"],
                        "attempts": len(action_result["schema_errors"]),
                    },
                    tool_result=None,
                    observation=action_result["error"],
                )
                turns.append(record)
                logger.append(record)
                if on_turn:
                    on_turn(record)
                snapshot_path = self._save_snapshot(
                    goal=goal,
                    context=context,
                    turns=turns,
                    runtime_state=runtime_state,
                    dangerous_tool_signatures=dangerous_sigs,
                )
                break
            action = action_result["action"]
            llm_raw_output = action_result["llm_raw_output"]
            schema_retry_count = len(action_result["schema_errors"])

            if action.action_type == "final_response":
                stop_reason = (
                    "goal_reached"
                    if self.stop_controller.check_goal_reached(action.content or "")
                    else "final_response"
                )
                record = TurnRecord(
                    turn=turn_idx,
                    goal=goal,
                    working_memory=working_memory,
                    llm_raw_output=llm_raw_output,
                    llm_action={**asdict(action), "schema_retry_count": schema_retry_count},
                    tool_result=None,
                    observation=action.content or "",
                )
                turns.append(record)
                logger.append(record)
                if on_turn:
                    on_turn(record)
                snapshot_path = self._save_snapshot(
                    goal=goal,
                    context=context,
                    turns=turns,
                    runtime_state=runtime_state,
                    dangerous_tool_signatures=dangerous_sigs,
                )
                return RunResult(
                    final_response=action.content or "",
                    turns=turns,
                    stop_reason=stop_reason,
                    log_path=str(logger.path),
                    snapshot_path=snapshot_path,
                )

            tool_result = self._execute_tool_call(
                tool_name=action.tool_name or "",
                arguments=action.arguments,
                dangerous_tool_signatures=dangerous_sigs,
                on_approve=on_approve,
            )
            runtime_state.budget_used += tool_result.attempts
            if not tool_result.ok:
                runtime_state.failure_count += 1

            observation = (
                f"tool={action.tool_name}; ok={tool_result.ok}; "
                f"output={tool_result.output}; error={tool_result.error}"
            )
            record = TurnRecord(
                turn=turn_idx,
                goal=goal,
                working_memory=working_memory,
                llm_raw_output=llm_raw_output,
                llm_action={**asdict(action), "schema_retry_count": schema_retry_count},
                tool_result=asdict(tool_result),
                observation=observation,
            )
            turns.append(record)
            logger.append(record)
            if on_turn:
                on_turn(record)
            self.memory.maybe_compress(turns)
            snapshot_path = self._save_snapshot(
                goal=goal,
                context=context,
                turns=turns,
                runtime_state=runtime_state,
                dangerous_tool_signatures=dangerous_sigs,
            )
            stop_reason = self.stop_controller.check_after_failure(runtime_state) or stop_reason
            if stop_reason == "max_failures_reached":
                break

        final_response = f"Stopped without final response. reason={stop_reason}"
        return RunResult(
            final_response=final_response,
            turns=turns,
            stop_reason=stop_reason,
            log_path=str(logger.path),
            snapshot_path=snapshot_path,
        )

    def resume(self, snapshot_path: str) -> RunResult:
        """Continue a previous run from a saved snapshot.

        Args:
            snapshot_path: Path to the snapshot JSON file produced by a prior run.

        Returns:
            A :class:`~harness.types.RunResult` continuing from the saved state.
        """
        return self.run(goal="", context=None, resume_from=snapshot_path)

    def _generate_action_with_schema_retry(self, working_memory: Dict[str, Any]) -> Dict[str, Any]:
        schema_errors: List[str] = []
        llm_outputs: List[Any] = []
        llm_input = dict(working_memory)

        for attempt in range(self.config.schema_retry_limit + 1):
            llm_raw_output = self.llm.generate(llm_input)
            llm_outputs.append(llm_raw_output)
            try:
                action = parse_llm_action(llm_raw_output)
                output_for_log: Any = llm_outputs[0] if len(llm_outputs) == 1 else llm_outputs
                return {
                    "ok": True,
                    "action": action,
                    "llm_raw_output": output_for_log,
                    "schema_errors": schema_errors,
                    "error": None,
                }
            except SchemaError as exc:
                schema_errors.append(str(exc))
                if attempt >= self.config.schema_retry_limit:
                    output_for_log = llm_outputs[0] if len(llm_outputs) == 1 else llm_outputs
                    return {
                        "ok": False,
                        "action": None,
                        "llm_raw_output": output_for_log,
                        "schema_errors": schema_errors,
                        "error": str(exc),
                    }
                llm_input = {
                    **working_memory,
                    "schema_feedback": {
                        "last_error": str(exc),
                        "required_types": ["tool_call", "final_response"],
                    },
                }

        return {
            "ok": False,
            "action": None,
            "llm_raw_output": llm_outputs,
            "schema_errors": schema_errors,
            "error": "Unexpected schema retry state.",
        }

    def _load_state(
        self,
        *,
        goal: str,
        context: Dict[str, Any],
        resume_from: str | None,
    ) -> Dict[str, Any]:
        if not resume_from:
            return {
                "goal": goal,
                "context": context,
                "turns": [],
                "summary": "",
                "failure_count": 0,
                "budget_used": 0,
                "dangerous_tool_signatures": [],
                "snapshot_path": None,
            }

        state = self.snapshot_store.load(resume_from)
        return {
            "goal": state["goal"],
            "context": state["context"],
            "turns": state["turns"],
            "summary": state.get("summary", ""),
            "failure_count": int(state.get("failure_count", 0)),
            "budget_used": int(state.get("budget_used", 0)),
            "dangerous_tool_signatures": list(state.get("dangerous_tool_signatures", [])),
            "snapshot_path": str(resume_from),
        }

    def _save_snapshot(
        self,
        *,
        goal: str,
        context: Dict[str, Any],
        turns: List[TurnRecord],
        runtime_state: StopState,
        dangerous_tool_signatures: List[str],
    ) -> str:
        return self.snapshot_store.save(
            {
                "goal": goal,
                "context": context,
                "turns": turns,
                "summary": self.memory.summary,
                "failure_count": runtime_state.failure_count,
                "budget_used": runtime_state.budget_used,
                "dangerous_tool_signatures": dangerous_tool_signatures,
            }
        )

    _APPROVAL_REQUIRED_TOOLS = frozenset(
        {
            "bash",
            "edit_file",
            "write_text_file",
            "write_file",
        }
    )

    def _needs_approval(self, tool_name: str) -> bool:
        """Determine if a tool call needs user approval based on trust level."""
        trust = self.config.trust_level
        if trust == "yolo":
            return False
        if trust == "auto-edit":
            return tool_name == "bash"
        # trust == "ask": all sensitive tools need approval
        return tool_name in self._APPROVAL_REQUIRED_TOOLS

    def _execute_tool_call(
        self,
        *,
        tool_name: str,
        arguments: Dict[str, Any],
        dangerous_tool_signatures: List[str],
        on_approve: Callable[[str, str, Dict[str, Any]], bool] | None = None,
    ) -> ToolExecutionResult:
        if self._needs_approval(tool_name):
            if on_approve is None:
                return ToolExecutionResult(
                    ok=False,
                    output=None,
                    error=(
                        f"Tool '{tool_name}' requires approval but no approval "
                        "callback provided. Use --trust yolo or run in "
                        "interactive mode."
                    ),
                )
            desc = self._describe_tool_call(tool_name, arguments)
            if not on_approve(tool_name, desc, arguments):
                return ToolExecutionResult(
                    ok=False,
                    output=None,
                    error="User denied tool execution",
                )

        fingerprint = json.dumps(
            {"tool_name": tool_name, "arguments": arguments}, sort_keys=True, ensure_ascii=False
        )
        if tool_name in self.config.dangerous_tools and fingerprint in dangerous_tool_signatures:
            return ToolExecutionResult(
                ok=False,
                output=None,
                error=f"Repeated dangerous tool call blocked: {tool_name}",
                blocked=True,
            )

        attempt = 0
        while True:
            attempt += 1
            result = self.tools.execute(tool_name, arguments)
            result.attempts = attempt
            if not self.error_policy.should_retry_tool(result, attempt):
                break

        if (
            tool_name in self.config.dangerous_tools
            and fingerprint not in dangerous_tool_signatures
        ):
            dangerous_tool_signatures.append(fingerprint)
        return result

    @staticmethod
    def _describe_tool_call(tool_name: str, arguments: Dict[str, Any]) -> str:
        if tool_name == "bash":
            return str(arguments.get("command", ""))
        if tool_name == "edit_file":
            path = arguments.get("path", "")
            old = str(arguments.get("old_text", ""))[:50]
            return f"{path}: {old!r}"
        if tool_name == "write_text_file":
            return str(arguments.get("path", ""))
        if tool_name == "write_file":
            return str(arguments.get("path", ""))
        if tool_name == "glob_files":
            return str(arguments.get("pattern", ""))
        if tool_name == "grep_search":
            return str(arguments.get("pattern", ""))
        if tool_name == "list_directory":
            return str(arguments.get("path", ""))
        return str(arguments)

File: ts/src/agent.ts
/**
 * HarnessAgent — single-agent execution harness.
 * Orchestrates the memory → LLM → tool/final → observe → next turn loop
 * with strict schema validation, reliability guardrails, and trajectory logging.
 */

import type {
  LLMAction,
  RunResult,
  ToolExecutionResult,
  TrustLevel,
  TurnRecord,
} from "./types.js";
import { toolError } from "./types.js";
import { SchemaError } from "./types.js";
import { parseLLMAction } from "./schema.js";
import { ToolDispatcher, registerCodingTools } from "./tools.js";
import { MemoryManager, type WorkingMemory } from "./memory.js";
import { ErrorPolicy } from "./error-policy.js";
import { SnapshotStore } from "./snapshot.js";
import { StopController, type StopState } from "./stop-controller.js";
import { TrajectoryLogger } from "./logger.js";
import type { BaseLLM } from "./llm.js";
import { SubAgentSpawner } from "./subagent.js";

/** Cast WorkingMemory to Record<string, unknown> for use with LLM/TurnRecord. */
function wmToRecord(wm: WorkingMemory): Record<string, unknown> {
  return wm as unknown as Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// RunConfig
// ---------------------------------------------------------------------------

export interface RunConfig {
  max_steps: number;
  log_dir: string;
  max_history_turns: number;
  schema_retry_limit: number;
  max_budget: number | null;
  max_failures: number | null;
  tool_retry_limit: number;
  snapshot_dir: string | null;
  dangerous_tools: string[];
  goal_reached_token: string | null;
  allowed_write_roots: string[];
  project_root: string;
  allow_bash: boolean;
  max_tokens_budget: number | null;
  trust_level: TrustLevel;
  agent_depth: number;
}

const DEFAULT_CONFIG: RunConfig = {
  max_steps: 20,
  log_dir: "logs",
  max_history_turns: 20,
  schema_retry_limit: 1,
  max_budget: null,
  max_failures: 3,
  tool_retry_limit: 0,
  snapshot_dir: null,
  dangerous_tools: [],
  goal_reached_token: null,
  allowed_write_roots: [],
  project_root: "",
  allow_bash: true,
  max_tokens_budget: null,
  trust_level: "ask",
  agent_depth: 0,
};

// ---------------------------------------------------------------------------
// Internal types
// ---------------------------------------------------------------------------

interface ActionResult {
  ok: boolean;
  action: LLMAction | null;
  llm_raw_output: unknown;
  schema_errors: string[];
  error: string | null;
}

interface RunOptions {
  resumeFrom?: string;
  onTurn?: (record: TurnRecord) => void;
  onApprove?: (toolName: string, description: string, args: Record<string, unknown>) => boolean;
  onToken?: (token: string) => void;
  onCompress?: () => void;
}

// ---------------------------------------------------------------------------
// HarnessAgent
// ---------------------------------------------------------------------------

export class HarnessAgent {
  readonly config: RunConfig;
  readonly tools: ToolDispatcher;
  readonly memory: MemoryManager;
  readonly errorPolicy: ErrorPolicy;
  readonly snapshotStore: SnapshotStore;
  readonly stopController: StopController;

  private static readonly _APPROVAL_REQUIRED_TOOLS = new Set([
    "bash",
    "edit_file",
    "write_text_file",
    "write_file",
    "save_memory",
    "spawn_agent",
  ]);

  constructor(
    public readonly llm: BaseLLM,
    config?: Partial<RunConfig>,
  ) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.tools = new ToolDispatcher({
      allowedWriteRoots: this.config.allowed_write_roots,
    });

    if (this.config.project_root) {
      registerCodingTools(this.tools, {
        allowBash: this.config.allow_bash,
        projectRoot: this.config.project_root,
      });
    }

    // Sub-agent integration: register spawn_agent when depth < 3
    if (this.config.agent_depth < 3) {
      const spawner = new SubAgentSpawner(this.config, this.llm, []);
      this.tools.registerTool(
        "spawn_agent",
        (args: Record<string, unknown>) => {
          // Note: this returns a Promise; async tool execution handled by caller
          return spawner.call(args);
        },
        {
          type: "object",
          description: "Spawn a child agent to accomplish a sub-goal",
          properties: {
            goal: {
              type: "string",
              description: "The goal for the child agent",
            },
            agent: {
              type: "string",
              description: "Optional agent definition name",
            },
            max_steps: {
              type: "integer",
              description: "Optional max steps override",
            },
          },
          required: ["goal"],
        },
      );
    }

    if (this.llm.setToolSchemas) {
      this.llm.setToolSchemas(this.tools.getToolSchemas());
    }

    this.memory = new MemoryManager(this.config.max_history_turns);
    this.errorPolicy = new ErrorPolicy(this.config.tool_retry_limit);
    this.snapshotStore = new SnapshotStore(
      this.config.snapshot_dir ?? this.config.log_dir,
    );
    this.stopController = new StopController(
      this.config.max_budget,
      this.config.max_failures,
      this.config.goal_reached_token,
    );
  }

  async run(
    goal: string,
    context?: Record<string, unknown> | null,
    options?: RunOptions,
  ): Promise<RunResult> {
    this.llm.onToken = options?.onToken;

    const state = this._loadState(goal, context ?? {}, options?.resumeFrom);
    const effectiveGoal = state.goal;
    const effectiveContext = state.context;
    const turns = state.turns;
    this.memory.summary = state.summary;
    const logger = new TrajectoryLogger(this.config.log_dir);
    let stopReason = "max_steps_reached";
    let snapshotPath: string | null = state.snapshotPath;
    const dangerousSigs: string[] = state.dangerousToolSignatures;
    const runtimeState: StopState = {
      budgetUsed: state.budgetUsed,
      failureCount: state.failureCount,
    };

    let totalTokens = 0;

    for (let step = 0; step < this.config.max_steps; step++) {
      stopReason =
        this.stopController.checkBeforeTurn(runtimeState) ?? stopReason;
      if (stopReason === "max_budget_reached") break;

      const turnIdx = turns.length + 1;
      const workingMemory = this.memory.buildWorkingMemory(
        effectiveGoal,
        effectiveContext,
        turns,
      );

      const actionResult = await this._generateActionWithSchemaRetry(
        wmToRecord(workingMemory),
      );

      // Track token usage
      const raw = actionResult.llm_raw_output;
      if (
        raw !== null &&
        typeof raw === "object" &&
        !Array.isArray(raw) &&
        "_usage" in (raw as Record<string, unknown>)
      ) {
        const usage = (raw as Record<string, unknown>)._usage;
        if (typeof usage === "object" && usage !== null) {
          totalTokens += ((usage as Record<string, unknown>).total_tokens as number) ?? 0;
        }
      }
      if (
        this.config.max_tokens_budget !== null &&
        totalTokens > this.config.max_tokens_budget
      ) {
        stopReason = "token_budget_exceeded";
        break;
      }

      runtimeState.budgetUsed += actionResult.schema_errors.length + 1;

      if (!actionResult.ok) {
        runtimeState.failureCount += 1;
        stopReason =
          this.stopController.checkAfterFailure(runtimeState) ?? "schema_error";
        const record: TurnRecord = {
          turn: turnIdx,
          goal: effectiveGoal,
          working_memory: wmToRecord(workingMemory),
          llm_raw_output: actionResult.llm_raw_output,
          llm_action: {
            type: "schema_error",
            error: actionResult.error,
            schema_errors: actionResult.schema_errors,
            attempts: actionResult.schema_errors.length,
          },
          tool_result: null,
          observation: actionResult.error ?? "",
        };
        turns.push(record);
        logger.append(record);
        if (options?.onTurn) options.onTurn(record);
        snapshotPath = this._saveSnapshot(
          effectiveGoal,
          effectiveContext,
          turns,
          runtimeState,
          dangerousSigs,
        );
        break;
      }

      const action = actionResult.action!;
      const llmRawOutput = actionResult.llm_raw_output;
      const schemaRetryCount = actionResult.schema_errors.length;

      if (action.action_type === "final_response") {
        stopReason = this.stopController.checkGoalReached(action.content ?? "")
          ? "goal_reached"
          : "final_response";

        const record: TurnRecord = {
          turn: turnIdx,
          goal: effectiveGoal,
          working_memory: wmToRecord(workingMemory),
          llm_raw_output: llmRawOutput,
          llm_action: {
            action_type: action.action_type,
            raw_output: action.raw_output,
            tool_name: action.tool_name,
            arguments: action.arguments,
            content: action.content,
            schema_retry_count: schemaRetryCount,
          },
          tool_result: null,
          observation: action.content ?? "",
        };
        turns.push(record);
        logger.append(record);
        if (options?.onTurn) options.onTurn(record);
        snapshotPath = this._saveSnapshot(
          effectiveGoal,
          effectiveContext,
          turns,
          runtimeState,
          dangerousSigs,
        );

        return {
          final_response: action.content ?? "",
          turns,
          stop_reason: stopReason,
          log_path: logger.path,
          snapshot_path: snapshotPath,
          total_tokens: totalTokens,
        };
      }

      // tool_call
      const toolResult = this._executeToolCall(
        action.tool_name ?? "",
        action.arguments,
        dangerousSigs,
        options?.onApprove,
      );
      runtimeState.budgetUsed += toolResult.attempts;
      if (!toolResult.ok) {
        runtimeState.failureCount += 1;
      }

      const observation =
        `tool=${action.tool_name}; ok=${toolResult.ok}; ` +
        `output=${toolResult.output}; error=${toolResult.error}`;

      const record: TurnRecord = {
        turn: turnIdx,
        goal: effectiveGoal,
        working_memory: wmToRecord(workingMemory),
        llm_raw_output: llmRawOutput,
        llm_action: {
          action_type: action.action_type,
          raw_output: action.raw_output,
          tool_name: action.tool_name,
          arguments: action.arguments,
          content: action.content,
          schema_retry_count: schemaRetryCount,
        },
        tool_result: {
          ok: toolResult.ok,
          output: toolResult.output,
          error: toolResult.error,
          retryable: toolResult.retryable,
          blocked: toolResult.blocked,
          attempts: toolResult.attempts,
        },
        observation,
      };
      turns.push(record);
      logger.append(record);
      if (options?.onTurn) options.onTurn(record);
      if (this.memory.maybeCompress(turns) && options?.onCompress) {
        options.onCompress();
      }
      snapshotPath = this._saveSnapshot(
        effectiveGoal,
        effectiveContext,
        turns,
        runtimeState,
        dangerousSigs,
      );
      stopReason =
        this.stopController.checkAfterFailure(runtimeState) ?? stopReason;
      if (stopReason === "max_failures_reached") break;
    }

    const finalResponse = `Stopped without final response. reason=${stopReason}`;
    return {
      final_response: finalResponse,
      turns,
      stop_reason: stopReason,
      log_path: logger.path,
      snapshot_path: snapshotPath,
      total_tokens: totalTokens,
    };
  }

  async resume(snapshotPath: string): Promise<RunResult> {
    return this.run("", null, { resumeFrom: snapshotPath });
  }

  // ---------------------------------------------------------------------------
  // Private methods
  // ---------------------------------------------------------------------------

  private async _generateActionWithSchemaRetry(
    workingMemory: Record<string, unknown>,
  ): Promise<ActionResult> {
    const schemaErrors: string[] = [];
    const llmOutputs: unknown[] = [];
    let llmInput: Record<string, unknown> = { ...workingMemory };

    for (
      let attempt = 0;
      attempt <= this.config.schema_retry_limit;
      attempt++
    ) {
      const llmRawOutput = await this.llm.generate(llmInput);
      llmOutputs.push(llmRawOutput);

      try {
        const action = parseLLMAction(llmRawOutput);
        const outputForLog =
          llmOutputs.length === 1 ? llmOutputs[0] : llmOutputs;
        return {
          ok: true,
          action,
          llm_raw_output: outputForLog,
          schema_errors: schemaErrors,
          error: null,
        };
      } catch (exc) {
        if (!(exc instanceof SchemaError)) throw exc;
        schemaErrors.push(exc.message);
        if (attempt >= this.config.schema_retry_limit) {
          const outputForLog =
            llmOutputs.length === 1 ? llmOutputs[0] : llmOutputs;
          return {
            ok: false,
            action: null,
            llm_raw_output: outputForLog,
            schema_errors: schemaErrors,
            error: exc.message,
          };
        }
        llmInput = {
          ...workingMemory,
          schema_feedback: {
            last_error: exc.message,
            required_types: ["tool_call", "final_response"],
          },
        };
      }
    }

    return {
      ok: false,
      action: null,
      llm_raw_output: llmOutputs,
      schema_errors: schemaErrors,
      error: "Unexpected schema retry state.",
    };
  }

  private _loadState(
    goal: string,
    context: Record<string, unknown>,
    resumeFrom?: string,
  ): {
    goal: string;
    context: Record<string, unknown>;
    turns: TurnRecord[];
    summary: string;
    failureCount: number;
    budgetUsed: number;
    dangerousToolSignatures: string[];
    snapshotPath: string | null;
  } {
    if (!resumeFrom) {
      return {
        goal,
        context,
        turns: [],
        summary: "",
        failureCount: 0,
        budgetUsed: 0,
        dangerousToolSignatures: [],
        snapshotPath: null,
      };
    }

    const state = this.snapshotStore.load(resumeFrom);
    return {
      goal: state.goal,
      context: state.context,
      turns: state.turns,
      summary: state.summary ?? "",
      failureCount: state.failure_count ?? 0,
      budgetUsed: state.budget_used ?? 0,
      dangerousToolSignatures: state.dangerous_tool_signatures ?? [],
      snapshotPath: resumeFrom,
    };
  }

  private _saveSnapshot(
    goal: string,
    context: Record<string, unknown>,
    turns: TurnRecord[],
    runtimeState: StopState,
    dangerousToolSignatures: string[],
  ): string {
    return this.snapshotStore.save({
      goal,
      context,
      turns,
      summary: this.memory.summary,
      failure_count: runtimeState.failureCount,
      budget_used: runtimeState.budgetUsed,
      dangerous_tool_signatures: dangerousToolSignatures,
    });
  }

  _needsApproval(toolName: string): boolean {
    const trust = this.config.trust_level;
    if (trust === "yolo") return false;
    if (trust === "auto-edit") return toolName === "bash";
    // trust === "ask": all sensitive tools need approval
    return HarnessAgent._APPROVAL_REQUIRED_TOOLS.has(toolName);
  }

  private _executeToolCall(
    toolName: string,
    args: Record<string, unknown>,
    dangerousToolSignatures: string[],
    onApprove?: (
      toolName: string,
      description: string,
      args: Record<string, unknown>,
    ) => boolean,
  ): ToolExecutionResult {
    // Approval check
    if (this._needsApproval(toolName)) {
      if (!onApprove) {
        return toolError(
          `Tool '${toolName}' requires approval but no approval ` +
            "callback provided. Use --trust yolo or run in interactive mode.",
        );
      }
      const desc = HarnessAgent._describeToolCall(toolName, args);
      if (!onApprove(toolName, desc, args)) {
        return toolError("User denied tool execution");
      }
    }

    // Dangerous tool fingerprint check
    const fingerprint = JSON.stringify(
      { tool_name: toolName, arguments: args },
      Object.keys({ tool_name: toolName, arguments: args }).sort(),
    );
    if (
      this.config.dangerous_tools.includes(toolName) &&
      dangerousToolSignatures.includes(fingerprint)
    ) {
      return toolError(
        `Repeated dangerous tool call blocked: ${toolName}`,
        { blocked: true },
      );
    }

    // Retry loop
    let attempt = 0;
    let result: ToolExecutionResult;
    do {
      attempt += 1;
      result = this.tools.execute(toolName, args);
      result.attempts = attempt;
    } while (this.errorPolicy.shouldRetryTool(result, attempt));

    // Cache dangerous fingerprint on success
    if (
      this.config.dangerous_tools.includes(toolName) &&
      !dangerousToolSignatures.includes(fingerprint)
    ) {
      dangerousToolSignatures.push(fingerprint);
    }

    return result;
  }

  static _describeToolCall(
    toolName: string,
    args: Record<string, unknown>,
  ): string {
    if (toolName === "bash") return String(args.command ?? "");
    if (toolName === "edit_file") {
      const filePath = args.path ?? "";
      const oldText = String(args.old_text ?? "").slice(0, 50);
      return `${filePath}: '${oldText}'`;
    }
    if (toolName === "write_text_file") return String(args.path ?? "");
    if (toolName === "write_file") return String(args.path ?? "");
    if (toolName === "spawn_agent") return String(args.goal ?? "");
    return JSON.stringify(args);
  }
}
